File: probes/src/v1/modules/layer.rs
use burn::{
    module::Module,
    nn::{Dropout, DropoutConfig, LayerNorm, LayerNormConfig},
    tensor::{Tensor, backend::Backend},
};

use crate::v1::modules::attention::{MhaInput, MultiHeadAttention, MultiHeadAttentionConfig};

use super::{
    encoder::EncoderConfig,
    pwff::{PositionWiseFeedForward, PositionWiseFeedForwardConfig},
};

#[derive(Module, Debug)]
pub struct Layer<B: Backend> {
    norm_1: LayerNorm<B>,
    attention: MultiHeadAttention<B>,
    norm_2: LayerNorm<B>,
    pwff: PositionWiseFeedForward<B>,
    dropout: Dropout,
}

impl<B: Backend> Layer<B> {
    pub(crate) fn new(config: &EncoderConfig, device: &B::Device) -> Self {
        Layer {
            norm_1: LayerNormConfig::new(config.embedding_dim).init(device),
            attention: MultiHeadAttentionConfig::new()
                .with_num_features(config.embedding_dim)
                .with_num_heads(config.num_heads)
                .with_dropout_rate(config.dropout_rate)
                .with_initializer(config.initializer.clone())
                .init(device),
            norm_2: LayerNormConfig::new(config.embedding_dim).init(device),
            pwff: PositionWiseFeedForwardConfig::new()
                .with_embedding_dim(config.embedding_dim)
                .with_hidden_dim(config.pwff_hidden_dim)
                .with_dropout_rate(config.dropout_rate)
                .with_initializer(config.initializer.clone())
                .init(device),
            dropout: DropoutConfig::new(config.dropout_rate).init(),
        }
    }
    pub(crate) fn forward(&self, context: Tensor<B, 3>) -> Tensor<B, 3> {
        let x = context;
        let residual_path = self.norm_1.forward(x.clone());
        let input_mha = MhaInput::new(residual_path);
        let residual_path = self.attention.forward(input_mha);
        let residual_path = self.dropout.forward(residual_path);
        let x = x + residual_path;
        let residual_path = self.norm_2.forward(x.clone());
        let residual_path = self.pwff.forward(residual_path);
        let residual_path = self.dropout.forward(residual_path);
        let x = x + residual_path;

        x
    }
}

File: python-models/src/soma_models/torch/v1/modules/layer.py
import torch
import torch.nn as nn

from soma_models.v1.configs import LayerConfig, PositionWiseFeedForwardConfig
from soma_models.torch.v1.modules.pwff import PositionWiseFeedForward
from soma_models.torch.v1.modules.attention import MultiHeadAttention


class Layer(nn.Module):
    def __init__(self, config: LayerConfig):
        super().__init__()
        self.norm_1 = nn.LayerNorm(config.embedding_dim, eps=1e-5)
        self.attention = MultiHeadAttention(
            num_heads=config.num_heads,
            num_features=config.embedding_dim,
            dropout_rate=config.dropout_rate,
            max_wavelength=config.max_wavelength,
            scale_factor=config.scale_factor,
        )
        self.norm_2 = nn.LayerNorm(config.embedding_dim, eps=1e-5)
        self.pwff = PositionWiseFeedForward(
            PositionWiseFeedForwardConfig(
                dropout_rate=config.dropout_rate,
                embedding_dim=config.embedding_dim,
                pwff_hidden_dim=config.pwff_hidden_dim,
            )
        )
        self.dropout = nn.Dropout(config.dropout_rate)

    def forward(self, context: torch.Tensor, positions: torch.Tensor, attn_mask: torch.Tensor):
        x = context
        residual_path = self.norm_1(x)
        residual_path = self.attention(residual_path, positions=positions, mask=attn_mask)
        residual_path = self.dropout(residual_path)
        x = x + residual_path
        residual_path = self.norm_2(x)
        residual_path = self.pwff(residual_path)
        residual_path = self.dropout(residual_path)
        x = x + residual_path
        return x
